test/tests/functional/pbs_soft_walltime.py

```diff
@@ -0,0 +1,686 @@
+
+        self.server.expect(JOB, {'Resource_List.soft_walltime': 5}, id=jid)
+
+    def test_soft_walltime_extend(self):
+        """
+        Test to see that soft_walltime is extended properly
+        """
+        J = Job(TEST_USER)
+        jid = self.server.submit(J)
+
+        self.server.alterjob(jid, {'Resource_List.soft_walltime': 5})
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid)
+
+        time.sleep(6)
+        self.server.manager(MGR_CMD_SET, SERVER, {'scheduling': 'True'})
+
+        self.server.expect(JOB, {'estimated.soft_walltime':
+                                 (MATCH_RE, '00:00:10|10')}, id=jid)
+
+        time.sleep(6)
+        self.server.manager(MGR_CMD_SET, SERVER, {'scheduling': 'True'})
+
+        self.server.expect(JOB, {'estimated.soft_walltime':
+                                 (MATCH_RE, '00:00:15|15')}, id=jid)
+
+    def test_soft_walltime_extend_hook(self):
+        """
+        Test to see that soft_walltime is extended properly when submitted
+        through a queue job hook
+        """
+        hook_body = \
+            """import pbs
+e = pbs.event()
+e.job.Resource_List["soft_walltime"] = 5
+e.accept()
+"""
+        a = {'event': 'queuejob', 'enabled': 'True'}
+        self.server.create_import_hook("que", a, hook_body)
+
+        J = Job(TEST_USER)
+        jid = self.server.submit(J)
+
+        self.server.expect(JOB, {'Resource_List.soft_walltime': 5}, id=jid)
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid)
+
+        time.sleep(6)
+        self.server.manager(MGR_CMD_SET, SERVER, {'scheduling': 'True'})
+
+        self.server.expect(JOB, {'estimated.soft_walltime':
+                                 (MATCH_RE, '00:00:10|10')}, id=jid)
+
+        time.sleep(6)
+        self.server.manager(MGR_CMD_SET, SERVER, {'scheduling': 'True'})
+
+        self.server.expect(JOB, {'estimated.soft_walltime':
+                                 (MATCH_RE, '00:00:15|15')}, id=jid)
+
+    def test_soft_then_hard(self):
+        """
+        Test to see if a job has both a soft and a hard walltime, that
+        the job's soft_walltime is not extended past its hard walltime.
+        It should first extend once and then extend to its hard walltime
+        """
+        J = Job(TEST_USER,
+                attrs={'Resource_List.ncpus': 1, 'Resource_List.walltime': 14})
+        jid = self.server.submit(J)
+
+        self.server.alterjob(jid, {'Resource_List.soft_walltime': 5})
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid)
+
+        time.sleep(6)
+        self.server.manager(MGR_CMD_SET, SERVER, {'scheduling': 'True'})
+
+        self.server.expect(JOB, {'estimated.soft_walltime':
+                                 (MATCH_RE, '00:00:10|10')}, id=jid)
+
+        time.sleep(6)
+        self.server.manager(MGR_CMD_SET, SERVER, {'scheduling': 'True'})
+
+        self.server.expect(JOB, {'estimated.soft_walltime':
+                                 (MATCH_RE, '00:00:14|14')}, id=jid)
```

Comment:
Looking at it again I am thinking there might be a race condition on line 348. we are checking after 12s and job is supposed to be killed by 14s. 2s (though it is enough) but might possibly introduce some race conditions. I think we should change the walltime to 17-18s.